src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/HPackEncoder.Dynamic.cs

```diff
@@ -0,0 +1,289 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace System.Net.Http.HPack
+{
+    internal partial class HPackEncoder
+    {
+        public const int DefaultHeaderTableSize = 4096;
+
+        // Internal for testing
+        internal readonly EncoderHeaderEntry Head;
+
+        private readonly bool _disableDynamicCompression;
+        private readonly EncoderHeaderEntry[] _headerBuckets;
+        private readonly byte _hashMask;
+        private uint _headerTableSize;
+        private uint _maxHeaderTableSize;
+        private bool _pendingTableSizeUpdate;
+        private EncoderHeaderEntry? _removed;
+
+        public HPackEncoder(bool disableDynamicCompression = false, uint maxHeaderTableSize = DefaultHeaderTableSize)
+        {
+            _disableDynamicCompression = disableDynamicCompression;
+            _maxHeaderTableSize = maxHeaderTableSize;
+            Head = new EncoderHeaderEntry();
+            Head.Initialize(-1, string.Empty, string.Empty, int.MaxValue, null);
+            // Bucket count balances memory usage and the expected low number of headers (constrained by the header table size).
+            // Performance with different bucket counts hasn't been measured in detail.
+            _headerBuckets = new EncoderHeaderEntry[16];
+            _hashMask = (byte)(_headerBuckets.Length - 1);
+            Head.Before = Head.After = Head;
+        }
+
+        public void UpdateMaxHeaderTableSize(uint maxHeaderTableSize)
+        {
+            if (_maxHeaderTableSize != maxHeaderTableSize)
+            {
+                _maxHeaderTableSize = maxHeaderTableSize;
+                _pendingTableSizeUpdate = true;
+
+                // Check capacity and remove entries that exceed the new capacity
+                EnsureCapacity(0);
+            }
+        }
+
+        public bool EnsureDynamicTableSizeUpdate(Span<byte> buffer, out int length)
+        {
+            if (_pendingTableSizeUpdate)
+            {
+                bool success = EncodeDynamicTableSizeUpdate((int)_maxHeaderTableSize, buffer, out length);
+                _pendingTableSizeUpdate = false;
+                return success;
+            }
+
+            length = 0;
+            return true;
+        }
+
+        public bool EncodeHeader(Span<byte> buffer, int staticTableIndex, HeaderEncodingHint encodingHint, string name, string value, out int bytesWritten)
+        {
+            Debug.Assert(!_pendingTableSizeUpdate, "Dynamic table size update should be encoded before headers.");
+
+            // Never index sensitive value.
+            if (encodingHint == HeaderEncodingHint.NeverIndex)
+            {
+                var index = ResolveDynamicTableIndex(staticTableIndex, name);
```

Comment:
Nit: we try to avoid using `var` when the type isn't obvious. I assume this is `int`?  Same number of characters, and clearer ;)